mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -738,21 +793,46 @@ LogicalResult ThreadwiseWriteAllRewritePattern::matchAndRewrite(
   Value bid = b.createOrFold<rock::WorkgroupIdOp>(loc, b.getIndexType());
   Value tid = b.createOrFold<rock::WorkitemIdOp>(loc, b.getIndexType());
 
-  SmallVector<Value, 3> writeStartCoords = {bid, tid, zero};
+  SmallVector<Value, 3> writeStartCoords{tid, zero};
+  SmallVector<int64_t, 3> bounds{1, numValues};
+  SmallVector<int64_t, 3> strides{1, vectorLen};
+  if (dstAddrSpace == gpu::AddressSpace::Global) {
+    writeStartCoords.insert(writeStartCoords.begin(), bid);
+    bounds.insert(bounds.begin(), 1);
+    strides.insert(strides.begin(), 1);
+  }
 
   auto outLoop = b.create<TransformingForOp>(
       loc, ArrayRef<ValueRange>{writeStartCoords, writeStartCoords},
-      ArrayRef<Attribute>{b.getArrayAttr({}), transforms},
-      ArrayRef<int64_t>{1, 1, numValues}, ArrayRef<int64_t>{1, 1, vectorLen},
+      ArrayRef<Attribute>{b.getArrayAttr({}), transforms}, bounds, strides,
       forceUnroll, useIndexDiffs);
   {
     OpBuilder::InsertionGuard guard(b);
     b.setInsertionPointToStart(outLoop.getBody());
-    b.create<GlobalStoreOp>(loc, source, buffer, b.getIndexAttr(vectorLen),
-                            op.getFeaturesAttr(), op.getStoreMethodAttr(),
-                            outLoop.getLowerCoords(/*domain=*/0)[2],
-                            outLoop.getValidity(/*domain=*/1),
-                            outLoop.getLowerCoords(/*domain=*/1));
+    if (dstAddrSpace == gpu::AddressSpace::Global) {
+      b.create<GlobalStoreOp>(
+          loc, source, buffer, b.getIndexAttr(vectorLen), op.getFeaturesAttr(),
+          op.getStoreMethodAttr(),
+          outLoop.getLowerCoords(/*domain=*/0)[iterDim[dstAddrSpace]],
+          outLoop.getValidity(/*domain=*/1),
+          outLoop.getLowerCoords(/*domain=*/1));
+    } else {
+      Type loadType = vectorTypeOrSelf(elementType, vectorLen);
+      TypedValue<IntegerType> valid = outLoop.getValidity(/*domain=*/0);
+      Value zeroInt = createZeroConstantOp(b, loc, valid.getType());
```

Comment:
This `cmp` is reduntant, - `validity` is already bool-typed

Reply:
right! fixed.